openedx/features/content_type_gating/tests/test_access.py

```diff
@@ -0,0 +1,98 @@
+
+    def setUp(self):
+        super(TestProblemTypeAccess, self).setUp()
+        self.course = CourseFactory.create(run='testcourse1', display_name='Test Course Title')
+        audit_mode = CourseModeFactory.create(course_id=self.course.id, mode_slug='audit')
+        verified_mode = CourseModeFactory.create(course_id=self.course.id, mode_slug='verified')
+        self.audit_user = UserFactory.create()
+        CourseEnrollmentFactory.create(user=self.audit_user, course_id=self.course.id, mode='audit')
+        with self.store.bulk_operations(self.course.id):
+            self.overview = ItemFactory.create(
+                parent=self.course,
+                display_name='Overview'
+            )
+            self.chapter_subsection = ItemFactory.create(
+                parent=self.overview,
+                category='sequential',
+                display_name='Lesson 1'
+            )
+            self.chapter_vertical = ItemFactory.create(
+                parent=self.chapter_subsection,
+                category='vertical',
+                display_name='Lesson 1 Vertical - Unit 1'
+            )
+
+    def is_block_gated(self, block, gated):
+        fake_request = Mock()
+        if gated:
+            # check that has_access raised the IncorrectPartitionGroupError in order to gate the block
+            with patch('courseware.access_response.IncorrectPartitionGroupError.__init__') as mock_access_error:
+                mock_access_error.return_value = IncorrectPartitionGroupError
+                with self.assertRaises(Http404):
+                    block = load_single_xblock(fake_request, self.audit_user.id, unicode(self.course.id),
+                                               unicode(block.scope_ids.usage_id), course=None)
+                self.assertTrue(mock_access_error.called)
+        else:
+            # check that has_access did not raise the IncorrectPartitionGroupError thereby not gating the block
+            with patch('courseware.access_response.IncorrectPartitionGroupError.__init__') as mock_access_error:
+                mock_access_error.return_value = IncorrectPartitionGroupError
+                block = load_single_xblock(fake_request, self.audit_user.id, unicode(self.course.id),
+                                           unicode(block.scope_ids.usage_id), course=None)
+                self.assertFalse(mock_access_error.called)
+
+    @ddt.data(
+        (False, False, 0, False),
+        (False, True, 0, False),
+        (False, False, 1, False),
+        (False, True, 1, False),
+        (True, False, 0, False),
+        (True, True, 0, False),
+        (True, False, 1, False),
+        (True, True, 1, True),
+    )
+    @ddt.unpack
+    def test_graded_score_weight_values(self, graded, has_score, weight, gated):
+        # has_score is determined by XBlock type. It is not a value set on an instance of an XBlock
+        if has_score:
+            block = ItemFactory.create(
```

Comment:
If we leave the code like this, then the course will be instantiated every time. Another option would be to use `SharedModuleStoreTestCase`, and explicitly create a graded problem and an ungraded problem inside the vertical, and then use `ddt` to pick between them.

Reply:
I guess that would require 4 problem blocks and 4 html blocks, which gets a bit clumsy. I'm not sure how much of a difference it would make, timing wise.

Reply:
perhaps the general course SetUp should use SharedModuleStoreTestCase but the problems themselves should be created in the test?